File: src/amr_controller/src/safety_arbitrator.cpp
#include <string>
#include <memory>
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/bool.hpp"

using std::placeholders::_1;

class SafetyArbitrator : public rclcpp::Node
{
public:
  SafetyArbitrator()
  : Node("safety_arbitrator_node"),
    lidar_stop_(false),
    ultrasonic_stop_(false),
    bumper_stop_(false)
  {
    declare_parameter<std::string>("lidar_safety_stop_topic",      "lidar_safety_stop");
    declare_parameter<std::string>("ultrasonic_safety_stop_topic", "ultrasonic_safety_stop");
    declare_parameter<std::string>("bumper_safety_stop_topic",     "bumper_safety_stop");
    declare_parameter<std::string>("safety_stop_topic",            "safety_stop");

    std::string lidar_topic      = get_parameter("lidar_safety_stop_topic").as_string();
    std::string ultrasonic_topic = get_parameter("ultrasonic_safety_stop_topic").as_string();
    std::string bumper_topic     = get_parameter("bumper_safety_stop_topic").as_string();
    std::string output_topic     = get_parameter("safety_stop_topic").as_string();

    lidar_sub_ = create_subscription<std_msgs::msg::Bool>(
      lidar_topic, 10,
      std::bind(&SafetyArbitrator::lidarCallback, this, _1));

    ultrasonic_sub_ = create_subscription<std_msgs::msg::Bool>(
      ultrasonic_topic, 10,
      std::bind(&SafetyArbitrator::ultrasonicCallback, this, _1));

    bumper_sub_ = create_subscription<std_msgs::msg::Bool>(
      bumper_topic, 10,
      std::bind(&SafetyArbitrator::bumperCallback, this, _1));

    safety_stop_pub_ = create_publisher<std_msgs::msg::Bool>(output_topic, 10);

    RCLCPP_INFO(get_logger(), "SafetyArbitrator started.");
    RCLCPP_INFO(get_logger(), "Listening to: %s, %s, %s",
      lidar_topic.c_str(), ultrasonic_topic.c_str(), bumper_topic.c_str());
    RCLCPP_INFO(get_logger(), "Publishing to: %s", output_topic.c_str());
  }

private:
  bool lidar_stop_;
  bool ultrasonic_stop_;
  bool bumper_stop_;

  rclcpp::Subscription<std_msgs::msg::Bool>::SharedPtr lidar_sub_;
  rclcpp::Subscription<std_msgs::msg::Bool>::SharedPtr ultrasonic_sub_;
  rclcpp::Subscription<std_msgs::msg::Bool>::SharedPtr bumper_sub_;
  rclcpp::Publisher<std_msgs::msg::Bool>::SharedPtr    safety_stop_pub_;

  void publishArbitratedStop()
  {
    bool combined = lidar_stop_ || ultrasonic_stop_ || bumper_stop_;
    std_msgs::msg::Bool msg;
    msg.data = combined;
    safety_stop_pub_->publish(msg);

    RCLCPP_INFO(get_logger(),
      "Safety arbitration -> %s  [lidar=%s | ultrasonic=%s | bumper=%s]",
      combined         ? "STOP"  : "CLEAR",
      lidar_stop_      ? "STOP"  : "clear",
      ultrasonic_stop_ ? "STOP"  : "clear",
      bumper_stop_     ? "STOP"  : "clear");
  }

  void lidarCallback(const std_msgs::msg::Bool & msg)
  {
    if (msg.data != lidar_stop_) {
      lidar_stop_ = msg.data;
      publishArbitratedStop();
    }
  }

  void ultrasonicCallback(const std_msgs::msg::Bool & msg)
  {
    if (msg.data != ultrasonic_stop_) {
      ultrasonic_stop_ = msg.data;
      publishArbitratedStop();
    }
  }

  void bumperCallback(const std_msgs::msg::Bool & msg)
  {
    if (msg.data != bumper_stop_) {
      bumper_stop_ = msg.data;
      publishArbitratedStop();
    }
  }
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<SafetyArbitrator>());
  rclcpp::shutdown();
  return 0;
}

File: src/amr_controller/amr_controller/safety_arbitrator.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import Bool


class SafetyArbitrator(Node):

    def __init__(self):
        super().__init__('safety_arbitrator_node')

        self.lidar_stop = False
        self.ultrasonic_stop = False
        self.bumper_stop = False

        # Declare parameters
        self.declare_parameter('lidar_safety_stop_topic', 'lidar_safety_stop')
        self.declare_parameter('ultrasonic_safety_stop_topic', 'ultrasonic_safety_stop')
        self.declare_parameter('bumper_safety_stop_topic', 'bumper_safety_stop')
        self.declare_parameter('safety_stop_topic', 'safety_stop')

        # Get parameters
        lidar_topic = self.get_parameter(
            'lidar_safety_stop_topic').get_parameter_value().string_value

        ultrasonic_topic = self.get_parameter(
            'ultrasonic_safety_stop_topic').get_parameter_value().string_value

        bumper_topic = self.get_parameter(
            'bumper_safety_stop_topic').get_parameter_value().string_value

        output_topic = self.get_parameter(
            'safety_stop_topic').get_parameter_value().string_value

        # Subscribers
        self.lidar_sub = self.create_subscription(
            Bool, lidar_topic, self.lidar_callback, 10)

        self.ultrasonic_sub = self.create_subscription(
            Bool, ultrasonic_topic, self.ultrasonic_callback, 10)

        self.bumper_sub = self.create_subscription(
            Bool, bumper_topic, self.bumper_callback, 10)

        # Publisher
        self.safety_stop_pub = self.create_publisher(Bool, output_topic, 10)

        self.get_logger().info("SafetyArbitrator started.")
        self.get_logger().info(
            f"Listening to: {lidar_topic}, {ultrasonic_topic}, {bumper_topic}")
        self.get_logger().info(f"Publishing to: {output_topic}")

    def publish_arbitrated_stop(self):
        combined = self.lidar_stop or self.ultrasonic_stop or self.bumper_stop

        msg = Bool()
        msg.data = combined
        self.safety_stop_pub.publish(msg)

        self.get_logger().info(
            f"Safety arbitration -> {'STOP' if combined else 'CLEAR'}  "
            f"[lidar={'STOP' if self.lidar_stop else 'clear'} | "
            f"ultrasonic={'STOP' if self.ultrasonic_stop else 'clear'} | "
            f"bumper={'STOP' if self.bumper_stop else 'clear'}]"
        )

    def lidar_callback(self, msg: Bool):
        if msg.data != self.lidar_stop:
            self.lidar_stop = msg.data
            self.publish_arbitrated_stop()

    def ultrasonic_callback(self, msg: Bool):
        if msg.data != self.ultrasonic_stop:
            self.ultrasonic_stop = msg.data
            self.publish_arbitrated_stop()

    def bumper_callback(self, msg: Bool):
        if msg.data != self.bumper_stop:
            self.bumper_stop = msg.data
            self.publish_arbitrated_stop()


def main(args=None):
    rclpy.init(args=args)
    node = SafetyArbitrator()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()